relay-kafka/src/producer/schemas.rs

```diff
@@ -0,0 +1,73 @@
+use std::fmt::Write;
+
+use jsonschema::JSONSchema;
+use thiserror::Error;
+
+use crate::config::{KafkaTopic, TopicAssignment, TopicAssignments};
+
+#[derive(Debug, Error)]
+pub enum SchemaError {
+    /// The "logical topic" is a concept in sentry-kafka-schemas and Snuba that identifies topics
+    /// irrespective of how they are structured in production. Snuba has a mapping from "logical
+    /// topic" to "physical topic" that is very similar to our `TopicAssignments`.
+    ///
+    /// The name of a logical topic is almost always equal to the Kafka topic's name in local
+    /// development. So, in order to determine a logical topic name from a `KafkaTopic`, we get
+    /// `TopicAssignments::default()` and try to resolve the name through that.
+    ///
+    /// When doing that we assume that Relay's default topic assignments don't use slicing/sharding
+    /// and no custom clusters.
+    ///
+    /// If somebody changes `impl Default for TopicAssignments` to have more complex defaults, this
+    /// error will start occurring. But it should not happen in prod.
+    #[error("failed to determine logical topic")]
+    InvalidLogicalTopic,
+
+    /// Failed to deserialize message, potentially because it isn't JSON?
+    #[error("failed to deserialize message")]
+    InvalidMessageJson(#[source] serde_json::Error),
+
+    /// Failed to deserialize schema as JSON
+    #[error("failed to deserialize schema")]
+    InvalidSchemaJson(#[source] serde_json::Error),
+
+    /// Failed to compile schema
+    // We stringify the inner error because `jsonschema::ValidationError` has weird lifetimes
+    #[error("failed to compile schema: {0}")]
+    InvalidSchemaCompiled(String),
+
+    /// Failed to validate message JSON against schema
+    // We stringify the inner error because `jsonschema::ValidationError` has weird lifetimes
+    #[error("message violates schema: {0}")]
+    InvalidMessage(String),
+}
+
+pub fn validate_message_schema(topic: KafkaTopic, message: &[u8]) -> Result<(), SchemaError> {
+    let default_assignments = TopicAssignments::default();
+    let logical_topic_name = match default_assignments.get(topic) {
+        TopicAssignment::Primary(logical_topic_name) => logical_topic_name,
+        _ => return Err(SchemaError::InvalidLogicalTopic),
+    };
+
+    let schema = match sentry_kafka_schemas::get_schema(logical_topic_name, None) {
```

Comment:
I know this is only for debug builds, but should we keep a cache of schemas by topic name so it isn't loaded from disk for every message? Maybe even put that cache in the `sentry-kafka-schemas` library?

Reply:
Maybe sometime but for now we don't want to be opinionated in how people use this library, so we settled for the simplest interface that comes with the least amount of dependencies (aka, return schema as string). This goes for the Python library as well: It requires you to bring your own schema validation library right now, and figure out yourself where it makes sense to store the schema. As such it doesn't have many dependencies.

I also don't want to build any cache right now because I foresee a (potential, not certain) future where we want/need people to iterate on schemas and relay in parallel (i.e. temporarily pointing relay to a local checkout of the schemas repo), and it would be good if changes to the schema files were picked up immediately in local development without recompiling.

I'm also not convinced that this schema validation library is the best one to use. I detest the fact that the schema and the payload has to be deserialized into intermediate `serde_json::Value`, but it seems that that is commonpractice for all schema validation libraries.

Reply:
> it would be good if changes to the schema files were picked up immediately in local development without recompiling.

I agree, I was thinking more of an in-memory cache that picks up the changes when you restart Relay (or re-run a test). I see your argument against putting such a cache in the library, but I still think it makes sense to put it _somewhere_ (e.g. the `KafkaClient`). I'm happy to contribute to this PR for this. Or do you see a use case for an actual hot reload, where you keep a local Relay instance running and edit the schema while sending in events?